src/member.rs

```diff
@@ -0,0 +1,478 @@
+
+use crate::crypto::{encryption, signing};
+use crate::id::{PublicId as TraitPublicId, SecretId};
+use crate::key_gen::message::Message;
+use crate::key_gen::outcome::Outcome;
+use crate::key_gen::{Error, KeyGen, Phase};
+use bincode::{deserialize, serialize};
+use bytes::Bytes;
+use crossbeam_channel::{unbounded, Receiver};
+use log::{info, trace};
+use quic_p2p::{Builder, Config, Event, Peer, QuicP2p, QuicP2pError, Token};
+use rand::{thread_rng, Rng};
+use schedule_recv::periodic_ms;
+use serde::{Deserialize, Serialize};
+use std::cmp::Ordering;
+use std::collections::{BTreeSet, HashMap};
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::thread::JoinHandle;
+use std::time::Duration;
+
+pub struct Member {
+    inner: Arc<Mutex<Inner>>,
+}
+
+impl Member {
+    pub fn new(config: Config) -> Result<Self, Error> {
+        let node = KeyInfo::new();
+
+        let (node_tx, node_rx) = unbounded::<Event>();
+        let (client_tx, _client_rx) = unbounded();
+
+        let quic_p2p = Builder::new(quic_p2p::EventSenders { node_tx, client_tx })
+            .with_config(config)
+            .build()
+            .map_err(|e| Error::QuicP2P(format!("{:#?}", e)))?;
+
+        let mut rng = thread_rng();
+
+        let inner = Inner {
+            quic_p2p,
+            id: rng.gen(),
+            group: Default::default(),
+            connected: Default::default(),
+            key_gen: None,
+            our_keys: node,
+        };
+        let arc_inner = Arc::new(Mutex::new(inner));
+
+        let _ = setup_quic_p2p_event_loop(&arc_inner, node_rx);
+
+        Ok(Self { inner: arc_inner })
+    }
+
+    #[cfg(test)]
+    pub fn new_for_test(config: Config) -> Result<(Self, SocketAddr), Error> {
+        let node = KeyInfo::new();
+
+        let (node_tx, node_rx) = unbounded::<Event>();
+        let (client_tx, _client_rx) = unbounded();
+
+        let mut quic_p2p = Builder::new(quic_p2p::EventSenders { node_tx, client_tx })
+            .with_config(config)
+            .build()
+            .map_err(|e| Error::QuicP2P(format!("{:#?}", e)))?;
+
+        let info = quic_p2p.our_connection_info().unwrap();
+        let mut rng = thread_rng();
+
+        let inner = Inner {
+            quic_p2p,
+            id: rng.gen(),
+            group: Default::default(),
+            connected: Default::default(),
+            key_gen: None,
+            our_keys: node,
+        };
+        let arc_inner = Arc::new(Mutex::new(inner));
+
+        let _ = setup_quic_p2p_event_loop(&arc_inner, node_rx);
+
+        Ok((Self { inner: arc_inner }, info))
+    }
+
+    pub fn init_dkg(
+        &mut self,
+        group: HashMap<NodeID, (SocketAddr, Peer)>,
+        threshold: usize,
+    ) -> Result<Message<NodeID>, Error> {
+        let mut inner_locked = self.inner.lock().unwrap();
+        inner_locked.group = group.clone();
+
+        // Maintain the list of conneted nodes
```

Comment:
We stay connected to our group until the session is done. And during network churns, if nodes gets dropped, `Member` would not know that nodes have dropped and would continue broadcasting messages pointlessly. Therefore we maintain this list of nodes, and it also ensures that we get connected to everybody before the initial broadcast as well.

Reply:
I am not sure here. I think for this crate, we are given the list of participants and not expected all to respond or got message delivered. 
For the situation of churning as you say, if there was too many, the DKG process just being block and report back who is blocking.

Reply:
But for initializing first. Do we not need to check if we are connected 100%. Because, if we do not wait and ensure connection to everybody, we might miss few initialization messages maybe?

Reply:
>for initializing first. Do we not need to check if we are connected 100%

we can carry out a check, however, I don't think it shall be a mandatory.

>we do not wait and ensure connection to everybody, we might miss few initialization messages maybe?

might be. However, that's what the real network will happen and the DKG process shall handle (tolerate with threshold or failed with blocking)